server/utextdocument.pas

```diff
@@ -355,29 +376,49 @@ procedure TextDocument_Completion(Rpc: TRpcPeer; Request: TRpcRequest);
     except
       on E: ERpcError do
       begin
-        // Unfortunately, there isn't really a good way to report errors to the
-        // client. While there are error responses, those aren't shown to the
-        // user. There is also the call window/showMessage, but this one is not
-        // implemented by NeoVim. So we work around it by showing a fake
-        // completion item.
         FreeAndNil(Response);
-        Response := TRpcResponse.Create(Request.Id);
-        Writer := Response.Writer;
-        Writer.Dict;
-          Writer.Key('items');
-          Writer.List;
-            Writer.Dict;
-              Writer.Key('label');
-              Writer.Str(e.Message);
-              Writer.Key('insertText');
-              Writer.Str('');
-            Writer.DictEnd;
-          Writer.ListEnd;
+        
+        if SyntaxErrorCausesLspError then
+        begin
+          Response := TRpcResponse.CreateError(Request.Id, 0, E.Message);
+        end else
+        begin
+          Response := TRpcResponse.Create(Request.Id);
+          Writer := Response.Writer;
+
+          Writer.Dict;
+            { Note that isIncomplete value is required.
+              See spec: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionList
+              Emacs actually throws Lisp errors when it is missing. }
+            Writer.Key('isIncomplete');
+            Writer.Bool(true);
+
+            Writer.Key('items');
+            Writer.List;
+
+              // Unfortunately, there isn't really a good way to report errors to the
+              // client. While there are error responses, those aren't shown to the
+              // user. There is also the call window/showMessage, but this one is not
+              // implemented by NeoVim. So we work around it by showing a fake
+              // completion item.
+              Writer.Dict;
+                Writer.Key('label');
+                Writer.Str(e.Message);
+                Writer.Key('insertText');
+                Writer.Str('');
+              Writer.DictEnd;
+
+            Writer.ListEnd;
+
+            //Writer.Key('activeParameter');
+            //Writer.Key('activeSignature');
+          Writer.DictEnd;
+        end;
 
-          //Writer.Key('activeParameter');
-          //Writer.Key('activeSignature');
-        Writer.DictEnd;
         Rpc.Send(Response);
+
+        if SyntaxErrorCausesShowMessage then
+          Rpc.Send(ShowErrorMessage(E.Message));
```

Comment:
Memory leak: The TRpcNotification created by `ShowErrorMessage` is not being freed.

I would also consider changing the type of `Response` to `TRpcMessageToClient` and then using it for both branches. But I'm also not really sure whether we need this new class despite the reasons you outlined in the comments. I think there should just add another constructor to `TRpcResponse` that creates an object without an Id. We already have:
```pascal
    constructor Create(Id: TRpcId);
    constructor CreateError(Id: TRpcId; Code: Integer; const Msg: string);
    constructor CreateRequest(const Method: string; Id: TRpcId);
```
I would just add another constructor:
```pascal
    constructor CreateMessage(const Msg: string);
```

Reply:
Fixed memory leak and I returned to 1 class just `TRpcResponse`. Yeah, this is just simpler :)

It exposes

```pascal
constructor CreateNotification(const Method: string);
```

instead of `CreateMessage` - I wanted to keep it generic for other LSP notifications in the future.